src/Security/PermissionSet/Depository/PermissionSet.php

```diff
@@ -0,0 +1,210 @@
+	{
+		$cdata = Contact::getPublicAndUserContactID($cid, $uid);
+		if (!empty($cdata)) {
+			$public_contact_str = $this->aclFormatter->toString($cdata['public']);
+			$user_contact_str   = $this->aclFormatter->toString($cdata['user']);
+			$cid                = $cdata['user'];
+		} else {
+			$public_contact_str = $this->aclFormatter->toString($cid);
+			$user_contact_str   = '';
+		}
+
+		$groups = [];
+		if (!empty($user_contact_str) && $this->db->exists('contact', [
+			'id' => $cid,
+			'uid' => $uid,
+			'blocked' => false
+		])) {
+			$groups = Group::getIdsByContactId($cid);
+		}
+
+		$group_str = '<<>>'; // should be impossible to match
+		foreach ($groups as $group_id) {
+			$group_str .= '|<' . preg_quote($group_id) . '>';
+		}
+
+		if (!empty($user_contact_str)) {
+			$condition = ["`uid` = ? AND (NOT (`deny_cid` REGEXP ? OR `deny_cid` REGEXP ? OR deny_gid REGEXP ?)
+				AND (allow_cid REGEXP ? OR allow_cid REGEXP ? OR allow_gid REGEXP ? OR (allow_cid = '' AND allow_gid = '')))",
+				$uid, $user_contact_str, $public_contact_str, $group_str,
+				$user_contact_str, $public_contact_str, $group_str];
+		} else {
+			$condition = ["`uid` = ? AND (NOT (`deny_cid` REGEXP ? OR deny_gid REGEXP ?)
+				AND (allow_cid REGEXP ? OR allow_gid REGEXP ? OR (allow_cid = '' AND allow_gid = '')))",
+				$uid, $public_contact_str, $group_str, $public_contact_str, $group_str];
+		}
+
+		return $this->select($condition);
+	}
+
+	/**
+	 * Fetch the default PermissionSet for a given user, create it if it doesn't exist
+	 *
+	 * @param int $uid
+	 *
+	 * @return Entity\PermissionSet
+	 * @throws Exception
+	 */
+	public function selectDefaultForUser(int $uid): Entity\PermissionSet
+	{
+		$self_contact = Contact::selectFirst(['id'], ['uid' => $uid, 'self' => true]);
+
+		return $this->selectOrCreate($this->factory->createFromString(
+			$uid,
+			$this->aclFormatter->toString($self_contact['id'])
+		));
+	}
+
+	/**
+	 * Fetch the empty PermissionSet for a given user, create it if it doesn't exist
+	 *
+	 * @param int $uid
+	 *
+	 * @return Entity\PermissionSet
+	 */
+	public function selectEmptyForUser(int $uid): Entity\PermissionSet
+	{
+		return $this->selectOrCreate($this->factory->createFromString($uid));
+	}
+
+	/**
+	 * Selects or creates a PermissionSet based on it's fields
+	 *
+	 * @param Entity\PermissionSet $permissionSet
+	 *
+	 * @return Entity\PermissionSet
+	 */
+	public function selectOrCreate(Entity\PermissionSet $permissionSet): Entity\PermissionSet
+	{
+		if ($permissionSet->id) {
+			return $permissionSet;
+		}
+
+		$fields = [
+			'uid'       => $permissionSet->uid,
+			'allow_cid' => $this->aclFormatter->toString($permissionSet->allow_cid),
+			'allow_gid' => $this->aclFormatter->toString($permissionSet->allow_gid),
+			'deny_cid'  => $this->aclFormatter->toString($permissionSet->deny_cid),
+			'deny_gid'  => $this->aclFormatter->toString($permissionSet->deny_gid),
+		];
```

Comment:
Please move this to a private method since it's used twice for the same purpose.